### ISSUE TITLE
Bad rotation interpolations

### ISSUE THREAD

**Issue by appgurueu** [minetest/minetest#10572]
Many rotations are poorly interpolated, including:

* Player look rotations (cinematic mode);
* Possibly: Entity & bone rotations

A poor interpolation uses a simple linear interpolation `current_rotation += (target_rotation - current_rotation) * interpolation`, which does not work well for angles, as it often ends up choosing a significantly longer interpolation. One example of a better interpolation is [Slerp](https://en.wikipedia.org/wiki/Slerp).

**Comment by hecktest** [comment]
Are bone rotations interpolated at all?

**Comment by appgurueu** [comment]
> Are bone rotations interpolated at all?

Probably not

**Comment by SmallJoker** [comment]
> which does not work well for angles, as it often ends up choosing a significantly longer interpolation

https://github.com/minetest/minetest/blob/f780bae05cc2fdd23a6d7326c770783da8d94ea3/src/util/numeric.h#L404-L421

**Comment by ElCeejo** [comment]
I thought I'd add to this since I played around with animating a object using set_bone_position a few days ago.

[Demonstration of using set_bone_position() to animate an object](https://youtu.be/3CWD7gXnUzE)

In the video obviously the rotation of bones is not smooth at all, especially when trying to move the bones quickly. Near the end you'll notice the "animation" freezes as well.

**Comment by Andrey2470T** [comment]
May I ask why do the bone rotations need the interpolation?

**Comment by appgurueu** [comment]
> May I ask just from a curiosity why do the bone rotations need the interpolation?

Visual smoothness

**Comment by appgurueu** [comment]
> > which does not work well for angles, as it often ends up choosing a significantly longer interpolation
> 
> This function was written to use interpolate in a closed/circular system. Other interpolations could be written using this as a base. ~If Irrlicht supports bone rotation intepolation, then~ it could also be used for that purpose.

No. The only proper way to interpolate rotations I currently see is a quaternion slerp one as implemented e.g. in https://github.com/minetest/minetest/pull/12388/ for bones. See e.g. [this wonderful demo of euler angle interpolation issues](https://splines.readthedocs.io/en/latest/rotation/naive-euler-angles-interpolation.html).
